src/vm/contracts.rs

```diff
@@ -1,57 +1,66 @@
 use vm::{SymbolicExpression, Value, apply, eval_all};
-use vm::errors::{Error, InterpreterResult as Result};
+use vm::errors::{Error, ErrType, InterpreterResult as Result, IncomparableError};
 use vm::callables::CallableType;
-use vm::contexts::{Environment, LocalContext, GlobalContext};
-use vm::database::{MemoryContractDatabase, ContractDatabase};
+use vm::contexts::{Environment, LocalContext, ContractContext, GlobalContext};
 use vm::parser;
-use vm::variables;
 
+#[derive(Serialize, Deserialize)]
 pub struct Contract {
-    db: Box<ContractDatabase>,
-    global_context: GlobalContext
+    contract_context: ContractContext,
 }
 
 impl Contract {
-    pub fn make_in_memory_contract(contract: &str) -> Result<Contract> {
+    pub fn initialize(name: &str, contract: &str, global_context: &mut GlobalContext) -> Result<Contract> {
         let parsed: Vec<_> = parser::parse(contract)?;
-        let mut global_context = GlobalContext::new();
-        let mut db_instance = Box::new(MemoryContractDatabase::new());
+        let mut contract_context = ContractContext::new(name.to_string());
 
-        let result = eval_all(&parsed, &mut *db_instance, &mut global_context)?;
+        let result = eval_all(&parsed, &mut contract_context, global_context)?;
         match result {
             Value::Void => {},
-            _ => return Err(Error::Generic("Contract instantiation should return null.".to_string()))
+            _ => return Err(Error::new(ErrType::InvalidArguments("Contract instantiation should return null.".to_string())))
         }
 
-        Ok(Contract { db: db_instance,
-                      global_context: global_context })
+        Ok(Contract { contract_context: contract_context })
     }
 
-    // Todo: add principal value type, check for sender to be one.
     pub fn execute_transaction(&mut self, sender: &Value, tx_name: &str,
-                               args: &[SymbolicExpression]) -> Result<Value> {
-        let func = self.global_context.lookup_function(tx_name)
-            .ok_or(Error::Undefined(format!("No such function in contract: {}", tx_name)))?;
+                               args: &[SymbolicExpression], global_context: &mut GlobalContext) -> Result<Value> {
+        let func = self.contract_context.lookup_function(tx_name)
+            .ok_or_else(|| { Error::new(ErrType::UndefinedFunction(tx_name.to_string())) })?;
         if !func.is_public() {
-            return Err(Error::Undefined("Attempt to call non-public function".to_string()))
+            return Err(Error::new(ErrType::NonPublicFunction(tx_name.to_string())));
         }
 
         if let Value::Principal(_, _) = sender {
```

Comment:
Can the situation arise where the `sender` _won't_ be a `Value::Principal`?

Reply:
My current inclination is "no, this shouldn't happen". Since this is something that the node itself should prevent from happening, I suppose this could be a panic rather than a normal error (somewhat similar to your SQLite error comments later on), however, I implemented it as an error for now --- the thinking is that what we probably want to do is differentiate two classes of errors in the VM: runtime, contract errors, and implementation errors. The implementation errors should eventually panic, but it's not clear if we want to have a higher level handler somewhere, or if we should just panic immediately at the point of failure.

Reply:
If this is the case -- that `sender` should never be anything but `Value::Principal`, then would it make more sense to update the method so that `sender` was always a `Value::Principal` variant, instead of just a `Value`?

Reply:
I could move the if statement higher up in the function, but I don't believe you can specify an enum variant as a method argument in Rust.